chat-client/src/client/withAdapter.ts

```diff
@@ -0,0 +1,151 @@
+import { MynahUI, MynahUIProps } from '@aws/mynah-ui'
+import { ChatClientAdapter, ChatEventHandler } from '../contracts/chatClientAdapter'
+import { disclaimerAcknowledgeButtonId } from './texts/disclaimer'
+
+type HandlerMethodName = keyof ChatEventHandler
+type HandlerParameters<T extends HandlerMethodName> = Parameters<NonNullable<ChatEventHandler[T]>>
+
+export const withAdapter = (
+    defaultEventHandler: ChatEventHandler,
+    mynahUIRef: { mynahUI: MynahUI | undefined },
+    chatClientAdapter: ChatClientAdapter
+): MynahUIProps => {
+    // Inject reference to MynahUI object into external event handler.
+    // This allows custom controllers to maintain drive Chat UI with custom, feature-specific logic.
+    const customEventHandler: ChatEventHandler = chatClientAdapter.createChatEventHandler(mynahUIRef)
+    if (!customEventHandler) {
+        throw new Error('Custom ChatEventHandler is not defined')
+    }
+
+    // Add default routing logic for event handlers with tabId as first argument.
+    const addDefaultRouting = (eventName: HandlerMethodName, defaultReturnValue?: any) => {
+        // @ts-ignore
+        return (...params) => {
+            // tabId is always the first argument
+            const tabId = params[0]
```

Comment:
maybe, we could add some validation for this: like check it's provided and `string`